### DISCUSSION TITLE
output.file audio segments made corrupted file

### DISCUSSION THREAD

**Discussion by cdgraff** [savonet/liquidsoap#2734] [User Support]
**Describe the bug**
Creating segments using output.file bring incorrectly created files with glitch in sound and errors FFmpeg report incorrect info inside the audio.

Tested with FFMPEG (Milliseconds of audio loose into transition from files and errors into archived files headers)
`    output.file(%ffmpeg(format="mp3", %audio(codec="libmp3lame",b="64k", samplerate=44100, channels=2)),`

And with native codec (here only missing MS of audio but no errores): 
```
   output.file(%mp3(bitrate=64, samplerate=44100, stereo=true),
       fallible=true,
       on_close=on_file_close,
       {"#{int_of_float(time())}.mp3"},reopen_delay=1.0,reopen_when={1s or 31s},cleanlive)

```

**To Reproduce**
source = input.http("your-icecast-url")

    #output.file(%ffmpeg(format="mp3", %audio(codec="libmp3lame",b="64k", samplerate=44100, channels=2)),
    output.file(%mp3(bitrate=64, samplerate=44100, stereo=true),
       fallible=true,
       on_close=on_file_close,
       {"#{int_of_float(time())}.mp3"},reopen_delay=1.0,reopen_when={1s or 31s},source)

**Expected behavior**
HLS Segmenter works perfect!

```
source = input.http("http:/your-icecast-source/test1")

# Define outputs formats

mp3 =
  %ffmpeg(
    #format="mpegts",
    format="mp3",
    %audio(
      channels=2,
      samplerate=44100,
      codec="libmp3lame",
      b="64k"
    )
  )

# Output to HLS
def segment_name(~position,~extname,stream_name) =
  timestamp = int_of_float(time())
  time.string("#{timestamp}.#{extname}")
end

def on_file_change(~state,fname) =
  if state == "closed" and file.extension(fname) != '.m3u8' then
    log.important(label="hls", "Segment #{fname} created")
  end
end

output.file.hls(id="output_hls",
                playlist="live.m3u8",
                fallible=true,
                segment_duration=30.0,
                segments=5,
                segments_overhead=5000,
                segment_name=segment_name,
                on_file_change=on_file_change,
                persist_at="state.config",
                !hlspath,
                [("mp3",mp3)],
                source)

```

**Version details**
Docker: savonet/liquidsoap:rolling-release-v2.1.x

**Install method**
Docker

**Comment by toots** [comment]
Looking at the code the HLS encoding doesn't do much for segmenting. The biggest difference is that it does not emit id3v2 tags. Have you tried adding `id3v2=false` to your encoder's parameters?

**Comment by cdgraff** [comment]
I tested with `drop_metadata(live)` and a few seconds ago, tested with `id3v2=false` and no errors, but still loosing data in the transition or almost listening to the files can hear the missing milliseconds in the transition.

After checking the waveform into the audio editor, the issue is at the beginning of the NEW file, each new file open has milliseconds of missing or noice data there

Trying to identify some difference, I found that 
`output.file` say Writing library  : Lavf59.27.100 (wrong result)

`output.file.hls` say Writing library : LAMEUUU (works perfect)

```
% mediainfo id3v2.mp3
General
Complete name                            : id3v2.mp3
Format                                   : MPEG Audio
File size                                : 705 KiB
Duration                                 : 1 min 30 s
Overall bit rate mode                    : Constant
Overall bit rate                         : 64.0 kb/s
Writing library                          : Lavf59.27.100

Audio
Format                                   : MPEG Audio
Format version                           : Version 1
Format profile                           : Layer 3
Format settings                          : Joint stereo / MS Stereo
Duration                                 : 1 min 30 s
Bit rate mode                            : Constant
Bit rate                                 : 64.0 kb/s
Channel(s)                               : 2 channels
Sampling rate                            : 44.1 kHz
Frame rate                               : 38.281 FPS (1152 SPF)
Compression mode                         : Lossy
Stream size                              : 705 KiB (100%)
```


```
% mediainfo /tmp/hls/2022/11/07/13h/hls.mp3
General
Complete name                            : /tmp/hls/2022/11/07/13h/hls.mp3
Format                                   : MPEG Audio
File size                                : 703 KiB
Duration                                 : 1 min 29 s
Overall bit rate mode                    : Constant
Overall bit rate                         : 64.0 kb/s
Writing library                          : LAMEUUU

Audio
Format                                   : MPEG Audio
Format version                           : Version 1
Format profile                           : Layer 3
Format settings                          : Joint stereo / MS Stereo
Duration                                 : 1 min 29 s
Bit rate mode                            : Constant
Bit rate                                 : 64.0 kb/s
Channel(s)                               : 2 channels
Sampling rate                            : 44.1 kHz
Frame rate                               : 38.281 FPS (1152 SPF)
Compression mode                         : Lossy
Stream size                              : 703 KiB (100%)
Writing library                          : LAMEUUU
```

Waveform from the transition, no data
<img width="1119" alt="image" src="https://user-images.githubusercontent.com/1087447/200365053-7f8c560a-26bd-48fa-8331-0e2608140ef1.png">

**Comment by toots** [comment]
The `ffmpeg` `mp3` format documentation is here: https://ffmpeg.org/ffmpeg-formats.html#mp3

You should try this:
```
output.file(%ffmpeg(
  format="mp3",
  id3v2_version=0,
  write_xing=0,
  %audio(
     codec="libmp3lame",
     b="64k",
    samplerate=44100,
    channels=2)
  ), ...)
```

Regarding the native `%mp3` encoder, would you be able to try with the versions that's building here: https://github.com/savonet/liquidsoap/actions/runs/3424388170 ?

It appears that there is another flush function for the encoder that we weren't using and that might yield better results.

**Comment by cdgraff** [comment]
Same issue in both cases: 
`Liquidsoap 2.1.3+git@a8460917`

Please listen at seconds 30 and 60 there is the merge, and no errors in the logs of FFmpeg, just listening when both audios are merged you can hear a gap in the sound.

Incorrect
FFMPEG test: 
[ffmpeg-file.mp3.zip](https://github.com/savonet/liquidsoap/files/9975287/ffmpeg-file.mp3.zip)

Native Test: 
[native-file.mp3.zip](https://github.com/savonet/liquidsoap/files/9975282/native-file.mp3.zip)

Correct:
HLS test:
[ffmpeg-hls.mp3.zip](https://github.com/savonet/liquidsoap/files/9975302/ffmpeg-hls.mp3.zip)

All the audios were captured in the same session.

Code used into the test:
```
source = input.http("http://icecast:port/test1")

# Define outputs formats

mp3 =
  %ffmpeg(
    format="mp3",
    id3v2_version=0,
    write_xing=0,
    %audio(
      channels=2,
      samplerate=44100,
      codec="libmp3lame",
      b="64k"
    )
  )

# Output to HLS
def segment_name(~position,~extname,stream_name) =
  timestamp = int_of_float(time())
  time.string("#{timestamp}.#{extname}")
end

def on_file_change(~state,fname) =
  if state == "closed" and file.extension(fname) != '.m3u8' then
    log.important(label="hls", "Segment #{fname} created")
  end
end

output.file.hls(id="output_hls",
                playlist="live.m3u8",
                fallible=true,
                segment_duration=30.0,
                segments=5,
                segments_overhead=5000,
                segment_name=segment_name,
                on_file_change=on_file_change,
                persist_at="state.config",
                "/tmp/hls/file-hls-ffmpeg",
                [("mp3",mp3)],
                source)

output.file(%mp3(bitrate=64, samplerate=44100, stereo=true, id3v2=false),
               fallible=true,
                #on_close=on_file_close,
                {"/tmp/hls/file-native/#{int_of_float(time())}.mp3"},reopen_delay=1.0,reopen_when={1s or 31s},source)

output.file(%ffmpeg(
  format="mp3",
  id3v2_version=0,
  write_xing=0,
  %audio(
     codec="libmp3lame",
     b="64k",
    samplerate=44100,
    channels=2)
  ),
  fallible=true,
  {"/tmp/hls/file-ffmpeg/#{int_of_float(time())}.mp3"},reopen_delay=1.0,reopen_when={1s or 31s},source)

```

**Comment by aferrarizype** [comment]
@toots some other test? I'm here to do all the research and test possible! I really need trust in the recording.

**Comment by toots** [comment]
I'm looking at it but, conceptually, it's a complex problem. The issues are:
* HLS is a gap-less, error-less sequence of media segments. However, the set of supported codecs is pretty limited which makes it easier.
* Some codecs require a global header, which complicates the HLS code a lot (see for instance the `mp4` specifications)
* In `output.file`, we do support all codecs and, so, when we re-open the file, we do flush and close the encoder and re-open a new file with a new encoder. This is likely to create audio artifact and prevent a proper concatenation of the resulting encoded files.

Given the specific of your needs, I'd suggest to implement a solution using the `output.file.hls` operator. The media resulting from this operator will always be guaranteed so be byte-exact and concatenable.

The operator provides a `on_file_change` callback, which will be called each time a new segment has finished being written (`state` argument will be `"closed"`). You should be able at this point to move or, better, copy the file to where you want your archive to be.

The only trick here would be to decide when to split. Segments are split based on how much has been encoded using `segment_duration`. Could that work for you?

**Comment by cdgraff** [comment]
thanks @toots, that was my plan B, just I was waiting to take a decision, and I agree with you on the problems that are facing with output.file

I'll implement output.file.hls, I'm just waiting v2.1.3 for the release of #2725 that I need to replicate the same path structure that I'm using currently with output.file.

Thanks for the great, support and time dedicated to troubleshooting that issue, and all your advices!

**Comment by toots** [comment]
Sounds good. I'll release `2.1.3` some time in december. The rolling release branch is aways intended to be release-ready at any time tho!
